share/rocm/cmake/header_template.h.in

```diff
@@ -8,7 +8,11 @@
 #ifndef ROCM_@ITEM_GUARD@_GAVE_WARNING
 #define ROCM_@ITEM_GUARD@_GAVE_WARNING
 
-#pragma message("This file is deprecated. Use the header file from @header_location@ by using #include <@correct_include@>")
+#if defined(_MSC_VER)
+#pragma message(": warning:This file is deprecated. Use the header file from @header_location@ by using #include <@correct_include@>")
+#elif defined(__GNUC__)
+#warning "This file is deprecated. Use the header file from @header_location@ by using #include <@correct_include@>"
+#endif
```

Comment:
Should there be a #else case as well, or remove the test for defined(__GNUC__)?
llvm does define __GNUC__, and this covers most of the compilers we use.

Reply:
No. Our public headers should be C99 compatible. It's fine to use extensions like `#pragma message` and `#warning`, but they should be guarded behind compiler-detection macros. These compiler checks more or less match [how rocSOLVER has handled deprecation warnings](https://github.com/ROCmSoftwarePlatform/rocSOLVER/blob/rocm-4.5.0/library/include/rocsolver-aliases.h#L21-L29) since ROCm 4.5.

Reply:
That being said... it's not exactly the same and our testing on this stuff is pretty sparse. I'm manually checking it right now.

Reply:
How do you write a perfect compiler detection macro? Until you can answer that you don't know if `#pragma GCC ` does what you think it does rather than launching nuclear missiles.

It is all a trade-off. Should we try and be helpful but with the risk we might get something wrong, or should we just stick to things we think we understand (but probably don't)?